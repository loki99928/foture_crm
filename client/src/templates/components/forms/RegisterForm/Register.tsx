import React, {useState, FC} from "react";
import cn from "classnames"
import {NavLink} from "react-router-dom";
import {useDispatch, useSelector} from "react-redux";
import {SubmitHandler, useForm} from "react-hook-form";
import * as Yup from "yup";

import bannerForm from "../../../assets/images/bg-head-form.jpg"
import {actionsAuth} from "../../../../redux/reducer/auth/actions";
import {getIsLoad, getMessage} from "../../../../redux/reducer/auth/selectors";
import s from "../util/Form.module.scss"
import Input from "../FomControls/Input";
import {FieldError} from "../FomControls/FeidError";
import {Button} from "../FomControls/Button";
import {fieldsForm} from "../util/FormType";
import {TUser} from "../../../../redux/reducer/auth/types";
import {yupResolver} from "@hookform/resolvers/yup";
import {validate} from "../util/Validate";
import {useClearForm} from "../../../../hooks/useClearForm";

const SignupSchema = Yup.object().shape({
    email: validate.email,
    password: validate.password
});

const Register: FC = () => {

    useClearForm()

    const {handleSubmit, formState: {errors, isValid}, ...handlers} = useForm<fieldsForm>({
        // defaultValues: {
        //     email: 'loki99928@yandex.ru',
        //     password: '123Qw@',
        // },
        mode: "all",
        resolver: yupResolver(SignupSchema)
    });

    const dispatch = useDispatch()
    const message = useSelector(getMessage)
    const isLoad = useSelector(getIsLoad)

    const onSubmit: SubmitHandler<fieldsForm> = (data: TUser) => {
        dispatch(actionsAuth.registerUserRequest(data))
    };

    return (
        <div className={s.blockForm}>
            <div className={s.formBanner}>
                <img src={bannerForm} alt=""/>
            </div>
            <div className={s.containerFields}>
                <h2 className={s.formTitle}>Registration Info</h2>
                <form onSubmit={handleSubmit(onSubmit)}>
                    <Input
                        errors={errors}
                        field='email'
                        handlers={handlers}
                        label='Your email'/>
                    <Input
                        errors={errors}
                        field='password'
                        handlers={handlers}
                        type='password'
                        label='Your password'/>
                    {message && <FieldError message={message}/>}
                    <Button disabled={isLoad || !isValid}/>
                </form>
                <div className={cn(s.formFooter, s.footer__form)}>
                    <NavLink to="/auth/">Authorize</NavLink>
                    <NavLink to="/forget/">forget password</NavLink>
                </div>
            </div>
        </div>
    )
}

export default Register
