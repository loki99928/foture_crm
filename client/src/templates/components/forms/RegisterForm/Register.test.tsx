import React from "react";
import {act, fireEvent, screen} from "@testing-library/react";

import Register from "./Register";
import {renderWithRouter} from "../../../../helpers/test/renderWithRouter";

describe('RegisterForm', () => {

    it('Snapshot', function () {
        const { asFragment } = renderWithRouter(<Register/>)
        expect(asFragment()).toMatchSnapshot()
    })

    describe('input email', () => {

        it('have', function () {
            renderWithRouter(<Register/>)
            expect(screen.queryByTestId('input_email')).toBeInTheDocument()
        })

        it('required', async () => {
            const { getByTestId } = renderWithRouter(<Register/>)
            const email = screen.getByTestId('input_email')
            await act( async () => {
                fireEvent.blur(email)
            });
            expect(getByTestId("formTextError")).not.toBe(null);
            expect(getByTestId("formTextError")).toHaveTextContent("Required");
        })

        it('validate', async () => {
            renderWithRouter(<Register/>)
            const email = screen.getByTestId('input_email')
            await act( async () => {
                fireEvent.change(email, {
                    target: { value: 'loki99928@yandex.ru' }
                })
            });
            await act( async () => {
                fireEvent.focusOut(email)
            });
            expect(screen.queryByTestId("formTextError")).toBeNull();
        })

        it('not validate', async () => {
            renderWithRouter(<Register/>)
            const email = screen.getByTestId('input_email')
            await act( async () => {
                fireEvent.change(email, {
                    target: { value: 'loki99928yandex' }
                })
            });
            await act( async () => {
                fireEvent.focusOut(email)
            });
            expect(screen.queryByTestId("formTextError")).not.toBeNull();
            expect(screen.queryByTestId("formTextError")).toHaveTextContent('Not a valid email');
        })

        it('too long', async () => {
            renderWithRouter(<Register/>)
            const email = screen.getByTestId('input_email')
            await act( async () => {
                fireEvent.change(email, {
                    target: { value: 'loki999282342234234234346ty546547456456546/jkdxjlmnkfvbcvb@yandex.ru' }
                })
            });
            await act( async () => {
                fireEvent.focusOut(email)
            });
            expect(screen.queryByTestId("formTextError")).not.toBeNull();
            expect(screen.queryByTestId("formTextError")).toHaveTextContent('Too Long!');
        })
    })

    describe('input password', () => {

        it('have',  () => {
            renderWithRouter(<Register/>)
            expect(screen.getByTestId('input_password')).toBeInTheDocument()
        })

        it('validate',  async () => {
            renderWithRouter(<Register/>)
            const password = screen.getByTestId('input_password')
            await act( async () => {
                fireEvent.change(password, {
                    target: { value: '123pass456345' }
                })
            });
            await act( async () => {
                fireEvent.focusOut(password)
            });
            expect(screen.queryByTestId("formTextError")).toBeNull();
        })

        it('too short', async () => {
            renderWithRouter(<Register/>)
            const password = screen.getByTestId('input_password')
            await act( async () => {
                fireEvent.change(password, {
                    target: { value: 'lk' }
                })
            });
            await act( async () => {
                fireEvent.focusOut(password)
            });
            expect(screen.queryByTestId("formTextError")).not.toBeNull();
            expect(screen.queryByTestId("formTextError")).toHaveTextContent('Too Short. Min length 5 symbol');
        })

        it('too long', async () => {
            renderWithRouter(<Register/>)
            const password = screen.getByTestId('input_password')
            await act( async () => {
                fireEvent.change(password, {
                    target: { value: '1232133554756786788566345345345345456756878989008907890678' }
                })
            });
            await act( async () => {
                fireEvent.focusOut(password)
            });
            expect(screen.queryByTestId("formTextError")).not.toBeNull();
            expect(screen.queryByTestId("formTextError")).toHaveTextContent('Too Long. Max length 50 symbol');
        })
    })

    describe('btn Send', () => {
        it('have btn Send', () => {
            renderWithRouter(<Register/>)
            expect(screen.getByRole('button', { name: /Send/i} )).toBeInTheDocument()
        })
        it('not disabled', async () => {
            const { getByTestId } = renderWithRouter(<Register/>)
            const btn = screen.getByRole('button', { name: /Send/i} )
            expect(btn).not.toBeDisabled()
        })
        it('disabled after error field of email', async () => {
            const { getByTestId } = renderWithRouter(<Register/>)
            const email = screen.getByTestId('input_email')
            const btn = screen.getByRole('button', { name: /Send/i} )
            await act( async () => {
                fireEvent.blur(email)
            });
            expect(btn).toBeDisabled()
        })
        it('not disabled after the error disappears', async () => {
            const { getByTestId } = renderWithRouter(<Register/>)
            const email = screen.getByTestId('input_email')
            const password = screen.getByTestId('input_password')
            const btn = screen.getByRole('button', { name: /Send/i} )
            await act( async () => {
                fireEvent.blur(email)
            });
            await act( async () => {
                fireEvent.change(email, {
                    target: { value: 'loki99928@yandex.ru' }
                })
            });
            await act( async () => {
                fireEvent.change(password, {
                    target: { value: '12356' }
                })
            });
            expect(btn).not.toBeDisabled()
        })
        it('disabled after error field of password', async () => {
            const { getByTestId } = renderWithRouter(<Register/>)
            const password = screen.getByTestId('input_password')
            const btn = screen.getByRole('button', { name: /Send/i} )
            await act( async () => {
                fireEvent.blur(password)
            });
            expect(btn).toBeDisabled()
        })
        it('clicking on a button with an empty form', async () => {
            renderWithRouter(<Register/>)
            const btn = screen.getByRole('button', { name: /Send/i} )
            await act(async () => {
                btn.click()
            })
            expect(screen.getAllByTestId('formTextError').length).toBe(2)
        })
    })

    // todo-dv нужно разобраться как тестировать переход по ссылкам
    describe('link', () => {
        describe('registration', () => {
            it('to have', () => {
                renderWithRouter(<Register/>)
                expect(screen.getByRole('link', {name: /Authorize/i}).closest('a')).toHaveAttribute('href', '/auth/')
            })
        })

        describe('forget', () => {
            it('to have',  () => {
                renderWithRouter(<Register/>)
                expect(screen.getByRole('link', {name: /forget password/i}).closest('a')).toHaveAttribute('href', '/forget/')
            })
        })
    })
})