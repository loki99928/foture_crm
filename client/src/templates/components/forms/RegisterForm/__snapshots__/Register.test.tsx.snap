// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegisterForm Snapshot 1`] = `
<DocumentFragment>
  <div
    class="blockForm"
  >
    <form
      action="#"
    >
      <div
        class="formBanner"
      >
        <img
          alt=""
          src="bg-head-form.jpg"
        />
      </div>
      <div
        class="containerFields"
      >
        <h2
          class="formTitle"
        >
          Registration Info
        </h2>
        <div
          class="formBlockField formBlock__field"
        >
          <input
            autocomplete="on"
            class="formField"
            data-testid="input_email"
            for="email"
            id="email"
            name="email"
            type="email"
            value=""
          />
          <div
            class="formLabel"
          >
            <label
              for="email"
            >
              Your email
            </label>
          </div>
        </div>
        <div
          class="formBlockField formBlock__field"
        >
          <input
            autocomplete="on"
            class="formField"
            data-testid="input_password"
            for="password"
            id="password"
            name="password"
            type="password"
            value=""
          />
          <div
            class="formLabel"
          >
            <label
              for="password"
            >
              Your password
            </label>
          </div>
        </div>
        <div>
          <button
            class="formBtn formBtn_hover formBtn_active"
            type="submit"
          >
            Send
          </button>
        </div>
        <div
          class="formFooter footer__form"
        >
          <a
            class=""
            href="/auth/"
          >
            Authorize
          </a>
          <a
            class=""
            href="/forget/"
          >
            forget password
          </a>
        </div>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
